### PR TITLE
Add Heel to Skeleton

### PR DESCRIPTION
**Pull request by tomer-algo**
Heels
- Init heel weights with ankle weights
- Add heel to joints mapping, PCKh calculations, etc.
- Add experiments for heel

Others
- Rotate image in inference according to baby-head
- Use get_weights_and_biases_owner() instead of specific hash

### CODE REVIEW

File: lib/dataset/JointsDataset.py [hunk 1]
```diff
@@ -323,3 +316,11 @@ def generate_target(self, joints, joints_vis):
...
 
         return target, target_weight
+
+    @staticmethod
+    def get_rotation_matrix_head_up(baby_bbox, head_bbox, img_height, img_width):
+        baby_center = baby_bbox.get_bbox_center()
+        head_center = head_bbox.get_bbox_center()
+        rotation_angle = np.rad2deg(-np.pi/2 - calc_vector_angle(baby_center, head_center))
+        rotation_matrix = cv2.getRotationMatrix2D((img_width/2, img_height/2), rotation_angle, 1.0)
+        return rotation_matrix
```
Comment (tomer-algo) [inline lib/dataset/JointsDataset.py:326]: assert for only rotation matirx (w/o scale, translation, etc.)

File: demo/demo_inference.py [hunk 2]
```diff
@@ -93,55 +95,54 @@ def do_predict(pose_model, image_path, baby_boxes, head_boxes, output_dir):
...
+    if rotate_head_up:
+        # Rotate image for skeleton detection
+        baby_box_obj = BBox(x1=baby_box[0][0], y1=baby_box[0][1], x2=baby_box[1][0], y2=baby_box[1][1])
+        head_box_obj = BBox(x1=head_box[0][0], y1=head_box[0][1], x2=head_box[1][0], y2=head_box[1][1])
+        image_height, image_width = image_pose.shape[:2]
+        rotation_matrix = JointsDataset.get_rotation_matrix_head_up(baby_box_obj, head_box_obj, image_height, image_width)
+        image_pose = cv2.warpAffine(image_pose, rotation_matrix, (image_width, image_height))
+
+        # Rotate baby box
+        points = np.array([[baby_box[0][0], baby_box[0][1]], [baby_box[1][0], baby_box[1][1]]]).reshape(2, 1, 2)
```
Comment (tomer-algo) [inline demo/demo_inference.py:116]: try to use baby_box_obj "coords to list"

File: demo/demo_inference.py [hunk 3]
```diff
@@ -93,55 +95,54 @@ def do_predict(pose_model, image_path, baby_boxes, head_boxes, output_dir):
...
-
-    #         baby_boxes[0][0] = tuple(cv2.transform(np.array([[baby_boxes[0][0]]]), M)[0][0])
-    #         baby_boxes[0][1] = tuple(cv2.transform(np.array([[baby_boxes[0][1]]]), M)[0][0])
+    if rotate_head_up:
+        # Rotate image for skeleton detection
+        baby_box_obj = BBox(x1=baby_box[0][0], y1=baby_box[0][1], x2=baby_box[1][0], y2=baby_box[1][1])
+        head_box_obj = BBox(x1=head_box[0][0], y1=head_box[0][1], x2=head_box[1][0], y2=head_box[1][1])
+        image_height, image_width = image_pose.shape[:2]
+        rotation_matrix = JointsDataset.get_rotation_matrix_head_up(baby_box_obj, head_box_obj, image_height, image_width)
+        image_pose = cv2.warpAffine(image_pose, rotation_matrix, (image_width, image_height))
```
Comment (tomer-algo) [inline demo/demo_inference.py:113]: check option to set output size as (max dim, max dim)
check baby & head bbox are in the rotated image
check also in training phase that the rotation make sense

File: lib/dataset/nanit_mpii.py [hunk 4]
```diff
@@ -256,6 +257,22 @@ def pckh_evaluate(self, preds):
 
         ]
         name_value = OrderedDict(name_value)
+
+        if 'Right_Heel' in self.joints_mapping and 'Left_Heel' in self.joints_mapping:
```
Comment (tomer-algo) [inline lib/dataset/nanit_mpii.py:261]: remove the sanity test if heels exists

File: lib/dataset/nanit_mpii.py [hunk 5]
```diff
@@ -256,6 +257,22 @@ def pckh_evaluate(self, preds):
...
+            rheel = self.joints_mapping['Right_Heel']
+            lheel = self.joints_mapping['Left_Heel']
+
+            # Add Heel to PCKh values
+            name_value['Heel'] = 0.5 * (PCKh[lheel] + PCKh[rheel])
+            name_value['Heel@0.1'] = 0.5 * (pckAll[rng_index_for_0_1, lheel] + pckAll[rng_index_for_0_1, rheel])
+
+            # Use Heel for MeanSmS@0.1
+            name_value['MeanSmS@0.1'] = 0.125 * (pckAll[rng_index_for_0_1, head] + pckAll[rng_index_for_0_1, neck] +
```
Comment (tomer-algo) [inline lib/dataset/nanit_mpii.py:270]: add to list and divide by list_len

File: lib/models/pose_hrnet.py [hunk 6]
```diff
@@ -490,11 +491,42 @@ def init_weights(self, pretrained=''):
                 if name.split('.')[0] in self.pretrained_layers \
                    or self.pretrained_layers[0] is '*':
                     need_init_state_dict[name] = m
+
+            self.handle_heel_init_if_needed(need_init_state_dict)
```
Comment (tomer-algo) [inline lib/models/pose_hrnet.py:495]: check option to create "pretrained" model with heel init (from ankle)

File: lib/dataset/JointsDataset.py (hunk 1 above)
Comment (tomer-algo) [reply lib/dataset/JointsDataset.py:326]: cv2.getRotationMatrix2D returns a rotation matrix

File: demo/demo_inference.py (hunk 2 above)
Comment (tomer-algo) [reply demo/demo_inference.py:116]: use new util function "transform_bbox"

File: demo/demo_inference.py (hunk 3 above)
Comment (tomer-algo) [reply demo/demo_inference.py:113]: doing rotation that way
- calc angle of baby-center to head-center -> rotation_angle
- calc rotation matrix: rotaion_angle where center is baby-center
- Add translation to move baby-center to image center

I did a little check
I trained (on ~1K images) with full rotation (not only 90 deg mul)
and then did inference on ~200 test images
1. with full rotation
2. with full rotation + uniform angle of -35 to 35 deg
option 1 had slightly better conf (~2% more) 

i’ll continue with option 1

File: lib/models/pose_hrnet.py (hunk 6 above)
Comment (tomer-algo) [reply lib/models/pose_hrnet.py:495]: saved model in "models/pytorch/pose_mpii/pose_hrnet_w32_256x256_with_heels.pth"

**Review by felixNan — APPROVED** [review]

File: lib/nanit_utils/utils_functions.py [hunk 7]
```diff
@@ -22,13 +22,23 @@ def translate_point(p, angle, translate_length):
     return p_trans
 
 def transform_box_xywh(box_xywh, transformation_matrix):
-    box_xywh_obj = gen_bbox_from_xywh_list(box_xywh)
-    box_xyxy = box_xywh_obj.coordinates_to_list('xyxy')
-    points = np.array([[box_xyxy[0], box_xyxy[1]], [box_xyxy[2], box_xyxy[3]]]).reshape(2, 1, 2)
-    box_xyxy_trans = cv2.transform(points, transformation_matrix)
-    box_transformed_obj = BBox(x1=box_xyxy_trans[0,0,0], y1=box_xyxy_trans[0,0,1], x2=box_xyxy_trans[1,0,0], y2=box_xyxy_trans[1,0,1])
+    box_obj = gen_bbox_from_xywh_list(box_xywh)
```
Comment (Nimrods) [inline lib/nanit_utils/utils_functions.py:25]: If I am not mistaken - there is already a ctor in BBox that does that

**Review by Nimrods — APPROVED** [review]
